upstairs/src/lib.rs

```diff
@@ -1555,44 +1538,86 @@ impl Downstairs {
                         // check for response encryption contexts here
                         if !response.encryption_contexts.is_empty() {
                             let mut successful_decryption = false;
+                            let mut successful_hash = false;
 
                             // Attempt decryption with each encryption context,
                             // and fail if all do not work. The most recent
                             // encryption context will most likely be the
                             // correct one so start there.
-                            let encryption_context_iter =
-                                response.encryption_contexts.iter().rev();
-                            for ctx in encryption_context_iter {
-                                // Note: decrypt_in_place does not overwrite
-                                // the buffer if it fails, otherwise we would
-                                // need to copy here.
-                                let decryption_result = context
-                                    .decrypt_in_place(
-                                        &mut response.data[..],
-                                        Nonce::from_slice(&ctx.nonce[..]),
-                                        Tag::from_slice(&ctx.tag[..]),
-                                    );
+                            let encryption_context_iter = response
+                                .encryption_contexts
+                                .iter()
+                                .enumerate()
+                                .rev();
+
+                            // Hashes and encryption contexts are written out at
+                            // the same time, therefore there should be the same
+                            // number of them.
+                            //
+                            // XXX shouldn't be an assert?
```

Comment:
It's a panic offence currently because both the EC and hash write occur in the same sqlite transaction, and the EC + hash truncation also occur in a transaction. I added an assert because based on that it categorizes as a server error.

Reply:
Removed XXX in 48a3f9f :)